constph.py

```diff
@@ -732,7 +733,98 @@ def setNumAttemptsPerUpdate(self, nattempts=None):
             # TODO: Perform enough titration attempts to ensure thorough mixing without taking too long per update.
             # TODO: Cache already-visited states to avoid recomputing?
             self.nattempts_per_update = self.getNumTitratableGroups()
-        
+
+
+class CalibrationTitration(MonteCarloTitration):
+    """Implementation of self-adjusted mixture sampling for calibrating titratable residues.
+
+
+
+    """
+    def __init__(self, system, temperature, pH, prmtop, cpin_filename, nattempts_per_update=None, simultaneous_proposal_probability=0.1, target_weights=None, debug=False):
+        """
+        target_weights (list) - None or nested list indexed[group][state] of relative weights (pi) for sams method
+        """
+
+        super(CalibrationTitration, self).__init__(system, temperature, pH, prmtop, cpin_filename, nattempts_per_update=None, simultaneous_proposal_probability=0.1, debug=False)
+
+        self.n_updates=0
+
+        for i,group in enumerate(self.titrationGroups):
+            for j, state in enumerate(self.titrationGroups[i]['titration_states']):
+                if target_weights is not None:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = target_weights[i][j]
+                else:
+                    self.titrationGroups[i]['titration_states'][j]['target_weight'] = 1.0 / len(self.titrationGroups[i]['titration_states'])
+
+    def update(self, context,scheme=None):
+        super(CalibrationTitration, self).update(context)
+        self.n_updates += 1
+
+        if scheme is not None:
+            temperature = self.temperature
+            kT = kB * temperature  # thermal energy
+            beta = 1.0 / kT  # inverse temperature
+            zeta = numpy.asarray(map(lambda x: x['relative_energy'], self.titrationGroups[0]['titration_states'][:]))
+            zeta *= -beta  # zeta^{t-1}
```

Comment:
Ooh, I wouldn't use unit-bearing quantities for free energies.  Those depend on temperature, and we may have a situation where multiple temperatures are in play, which will break everything.

Just use unitless free energies (implicit units of kT) everywhere.  Just thing of them as negative log probabilities.